tools/smoke_test.py

```diff
@@ -0,0 +1,166 @@
+            raise RuntimeError(
+                "torchdynamo requirements not met:\n"
+                + "\n".join(map(str, dynamo_broken_deps))
+            )
+    proc = subprocess.run(["pip", "show", "torchdynamo"], capture_output=True)
+    if proc.returncode != 0:
+        raise RuntimeError("`torchdynamo` is not installed")
+
+
+# Checks for correct torch version.
+# Using check_pip_deps does not work if the minimum required torch version
+# is not present in PyPI.
+def check_torch():
+    import torch
+
+    if torch.__version__ < MIN_TORCH_VERSION:
+        raise RuntimeError(
+            f"torch version not supported: {torch.__version__} "
+            f"- minimum requirement: {MIN_TORCH_VERSION}"
+        )
+
+    return torch.__version__
+
+
+def check_cuda():
+    import torch
+
+    if not torch.cuda.is_available():
+        return None
+
+    torch_cuda_ver = Version(torch.version.cuda)
+    if torch_cuda_ver < MIN_CUDA_VERSION:
+        raise RuntimeError(
+            f"CUDA version not supported: {torch_cuda_ver} "
+            f"- minimum requirement: {MIN_CUDA_VERSION}"
+        )
+
+    # check if torch cuda version matches system cuda version
+    cuda_home = os.getenv("CUDA_HOME")
+    if not cuda_home:
+        raise RuntimeError("$CUDA_HOME not found")
+
+    nvcc = os.path.join(cuda_home, "bin", "nvcc")
+    nvcc_version_str = (
+        subprocess.check_output([nvcc, "--version"]).strip().decode("utf-8")
+    )
+    cuda_version = re.search(r"release (\d+[.]\d+)", nvcc_version_str)
+    if cuda_version is None:
+        raise RuntimeError("CUDA version not found in `nvcc --version` output")
+
+    cuda_ver = Version(cuda_version.group(1))
+    if cuda_ver != torch_cuda_ver:
+        raise RuntimeError(
+            f"CUDA version mismatch, torch version: {torch_cuda_ver}, env version: {cuda_ver}"
+        )
+
+    return cuda_ver
+
+
+def check_dynamo(backend, device, err_msg):
+    import torch
+
+    if device == "cuda" and not torch.cuda.is_available():
+        print(f"CUDA not available -- skipping CUDA check on {backend} backend\n")
+        return
+
+    try:
+        import torchdynamo
+
+        torchdynamo.reset()
+
+        @torchdynamo.optimize(backend, nopython=True)
+        def fn(x):
+            return x + x
+
+        class Module(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+
+            def forward(self, x):
+                return x + x
+
+        mod = Module()
+        opt_mod = torchdynamo.optimize(backend, nopython=True)(mod)
+
+        for f in (fn, opt_mod):
+            x = torch.randn(10, 10).to(device)
+            x.requires_grad = True
+            y = f(x)
+            torch.testing.assert_close(y, x + x)
+            z = y.sum()
+            z.backward()
+            torch.testing.assert_close(x.grad, 2 * torch.ones_like(x))
+    except Exception:
+        sys.stderr.write(traceback.format_exc() + "\n" + err_msg + "\n\n")
+        sys.exit(1)
```

Comment:
This just feels like a unit test snuck into this - what is the point of this? If dynamo is installed, it should be working. If its installed, running any unit test should work. This just makes things confusing. If you want to keep the spirit of a dynamo check of this complexity here, just run pytest on test_functions.py or something light.

Reply:
I think this is relevant to user guide vs. developer guide. The purpose of this smoke test is for user guide, we should not expect users are dynamo expert like us and should not expose unit tests to them.

Reply:
This is a unit test exposed to them in all but name.

Reply:
I disagree but do not care enough to argue lol.

Reply:
No worry, we can revisit if the requirement from Peng/Jason/Mark is not reasonable.

Reply:
I'm inclined to keep a simple test here since (1) it's difficult for incorrect triton versions to be detected otherwise, and (2) pytest and other testing libraries should not be required dependencies for users.